OEL-2674: Implement email validation for the subscription link

AaronGilMartinez opened the pull request.

AaronGilMartinez committed:
OEL-2674: Add schema.

AaronGilMartinez committed:
OEL-2674: Add decoupled auth module.

AaronGilMartinez committed:
OEL-2674: Add service to manage anonymous subscriptions.

AaronGilMartinez committed:
OEL-2674: Change parameter to FlagInterface.

AaronGilMartinez committed:
OEL-2541: Add interface.

AaronGilMartinez committed:
OEL-2674: Add routes, controller methods and form submission.

AaronGilMartinez committed:
OEL-2674: Add more asserts.

AaronGilMartinez committed:
OEL-2674: Replace PHP functions with Drupal API.

AaronGilMartinez committed:
OEL-2674: Fix multiple issues like routing, controller, keep subscriptions, rename test, change URL to link and add form test.

AaronGilMartinez committed:
OEL-2674: REmove unneeded function.

AaronGilMartinez committed:
OEL-2674: Add tests for subscription, messaging for error, and manager tests.

AaronGilMartinez committed:
OEL-2541: Remove unnecessary create and entity link.

AaronGilMartinez committed:
OEL-2674: Fix typo.

AaronGilMartinez committed:
OEL-2674: Change functions description.

AaronGilMartinez committed:
OEL-2674: Test comments and remove unneeded execution.

AaronGilMartinez committed:
OEL-2674: Test comments and remove unneeded execution.

AaronGilMartinez committed:
OEL-2674: Add changed field, and expired hash control.

AaronGilMartinez committed:
OEL-2674: Add more expired hash tests.

AaronGilMartinez committed:
OEL-2674: Fix comments and change return.

brummbar commented on modules/oe_subscriptions_anonymous/oe_subscriptions_anonymous.install:
```diff
@@ -0,0 +1,58 @@
+    'fields' => [
+      'mail' => [
+        'description' => 'Subscription mail.',
+        'length' => 255,
```
Let's use the same value as \Drupal\Core\Field\Plugin\Field\FieldType\EmailItem::schema .

AaronGilMartinez replied:
Added

brummbar commented on modules/oe_subscriptions_anonymous/oe_subscriptions_anonymous.install:
```diff
@@ -0,0 +1,58 @@
+        'length' => 255,
+        'type' => 'varchar_ascii',
+        'not null' => TRUE,
+        'default' => '',
```
Is there a reason for empty string as default?

AaronGilMartinez replied:
Removed

brummbar commented on modules/oe_subscriptions_anonymous/oe_subscriptions_anonymous.install:
```diff
@@ -0,0 +1,58 @@
+      ],
+      'changed' => [
+        'description' => 'Last time the subscription was changed.',
+        'length' => 255,
```
Int is the correct type.

AaronGilMartinez replied:
Changed

brummbar commented on modules/oe_subscriptions_anonymous/oe_subscriptions_anonymous.routing.yml:
```diff
@@ -6,3 +6,19 @@ oe_subscriptions_anonymous.anonymous_subscribe:
     _anonymous_subscription_access: 'TRUE'
+
+oe_subscriptions_anonymous.anonymous_confirm:
+  path: '/confirm/{flag}/{entity_id}/{email}/{hash}'
```
These routes are too generic, we might clash with other modules. We need to pick a starting point for these.
Let's have them start with /subscribe/, so /subscribe/confirm, subscribe/cancel, subscribe/request

AaronGilMartinez replied:
Added

brummbar commented on modules/oe_subscriptions_anonymous/src/AnonymousSubscriptionManager.php:
```diff
@@ -0,0 +1,248 @@
+   */
+  public function createSubscription(string $mail, FlagInterface $flag, string $entity_id): string {
+    // Validate mail.
+    if (!$this->mailValidatorService->isValid($mail)) {
```
It's not the job of this service to validate the email.

AaronGilMartinez replied:
Removed

brummbar commented on modules/oe_subscriptions_anonymous/src/AnonymousSubscriptionManager.php:
```diff
@@ -0,0 +1,248 @@
+/**
+ * Class to manage anonymous subscriptions.
+ */
+class AnonymousSubscriptionManager implements AnonymousSubscriptionManagerInterface {
```
This needs to be split in two services.
One service will handle tokens. We will need this later also for allowing anon users to access their subscriptions page. It should have methods like:

1. `isValid(string $token, string $email, string $scope)`: checks if a token exists and is valid (not expired).
2. `get(string $email, string $scope)`: returns a token. It will create the entry if needed, or updated it.
3. `delete(string $email, string $scope)`: deletes a token given some parameters.
4. `deleteExpired()`: deletes all expired tokens.

For `$scope` above, we need to choose what to do:
- we have a single service for the subscription process and the subscriptions edit page. In this case, we build the scope using a static method in an helper class.
- we have a service per token type, but we need to have double interfaces and all.

brummbar replied:
The other service will be the AnonymousSubscriptionManager, which should just take care of subscribing anonymous users. The method for now should be just `subscribe(string $email, FlagInterface $flag, string $entity_id)`.
The validation of the hash is to be done in the controller (like core does).

AaronGilMartinez replied:
Service splitted, using a single service for the scope.

brummbar commented on modules/oe_subscriptions_anonymous/src/AnonymousSubscriptionManager.php:
```diff
@@ -0,0 +1,248 @@
+
+    // Create decoupled user.
+    if ($account === FALSE) {
+      $user = User::create(['mail' => $mail])->save();
+      $account = User::load($user);
```
Always use dependency injection.

AaronGilMartinez replied:
Done

brummbar reviewed (changes requested):
Let's start with these changes.

AaronGilMartinez committed:
OEL-2674: Use mail max length for schema field.

AaronGilMartinez committed:
OEL-2674: Remove empty default falues in schema.

AaronGilMartinez committed:
OEL-2674: Change field type.

AaronGilMartinez committed:
OEL-2674: Add routes prefix.

AaronGilMartinez committed:
OEL-2674: Remove mail validation in service.

AaronGilMartinez committed:
OEL-2674: Add entity type manager for creating users.

AaronGilMartinez committed:
OEL-2674: Split subscrition manager service, add scope.

AaronGilMartinez committed:
OEL-2674: Comments fixes.

brummbar committed:
OEL-2674: Reflect the scope of the table in the name.

brummbar committed:
OEL-2674: Rename service to reflect its scope.

This service doesn't store subscriptions, it stores tokens
that will be used to validate an email in order to fullfill
an action (subscription, view, etc.).

brummbar committed:
OEL-2674: Always use service for time.

brummbar committed:
OEL-2674: Count queries to check if results exists.

brummbar committed:
OEL-2674: Validate when the email is assigned to a coupled user.

brummbar committed:
OEL-2674: Prepare the email in the dedicated hook.

This allows to send the email from other pages, e.g. an administration
page where the website admin can resend the email. Keeping the code
in the form submit limits the reusability.

brummbar committed:
OEL-2674: Cleanup subscribe method.

brummbar committed:
OEL-2674: Canceling active subscriptions should be done only from the user settings.

If a user is already registered, we should send a different e-mail with an unsubscribe link.

brummbar committed:
OEL-2674: Add a base kernel test.

brummbar committed:
OEL-2674: Use a separate test per service.

brummbar committed:
OEL-2674: Cleanup interface.

brummbar committed:
OEL-2674: Rework token manager test.

brummbar committed:
OEL-2674: Rework subscription manager test.

brummbar committed:
OEL-2674: Adapt code after rebase.

brummbar committed:
OEL-2674: Add mail testing to existing tests.

brummbar committed:
OEL-2674: Improve mail subject and messages.

brummbar committed:
OEL-2674: Test the email sent.

brummbar committed:
OEL-2674: Refactor the code to reduce code duplication.

brummbar committed:
OEL-2674: Test cancel link, uniform messages to be warnings when links are not valid.

brummbar committed:
OEL-2674: Drop testdox.

brummbar committed:
OEL-2674: Rename test classes.

brummbar committed:
OEL-2674: Test the redirect.

brummbar committed:
OEL-2674: Add test coverage for invalid and expired tokens for the subscription request routes.

brummbar committed:
OEL-2674: Rework test to cover all the routes with the custom access check.

brummbar committed:
OEL-2674: Coding standards.

brummbar committed:
OEL-2674: Improve route naming.
